File: GameFramePro/Assets/Scripts/Core/Base/Module/ResourcesManager/AssetBundle_Load/LoadAssetBundle/AssetBundleSubAssetLoadRecord.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System;


namespace GameFramePro.ResourcesEx
{
    /// <summary>
    /// AssetBundle 包中加载的子资源
    /// </summary>
#if UNITY_EDITOR
    [System.Serializable]
#endif
    public class AssetBundleSubAssetLoadRecord : ILoadAssetRecord
    {

#if UNITY_EDITOR
        #region Show
        public string Debug_AssetUrl = string.Empty;
        public int Debug_InstanceID = 0;
        public int Debug_ReferenceCount = 1;
        public LoadedAssetTypeEnum Debug_AssetLoadedType = LoadedAssetTypeEnum.None;
        public UnityEngine.Object Debug_TargetAsset = null;
        public float Debug_RemainTimeToBeDelete = 0;
        public long Debug_MarkToDeleteTime = 0;
        public IAssetManager Debug_BelongAssetManager = null;
        public string Debug_AssetBelongBundleName;


        public void UpdateData()
        {
            Debug_InstanceID = InstanceID;
            Debug_AssetUrl = AssetUrl;
            Debug_ReferenceCount = ReferenceCount;
            Debug_AssetLoadedType = AssetLoadedType;
            Debug_TargetAsset = TargetAsset;
            Debug_RemainTimeToBeDelete = RemainTimeToBeDelete;
            Debug_MarkToDeleteTime = MarkToDeleteTime;
            Debug_BelongAssetManager = BelongAssetManager;
            Debug_AssetBelongBundleName = AssetBelongBundleName;
        }

        #endregion
#endif

        public int InstanceID { get; protected set; } = 0;
        public string AssetUrl { get; protected set; } = string.Empty;
        public int ReferenceCount { get; protected set; } = 0;
        public LoadedAssetTypeEnum AssetLoadedType { get; protected set; } = LoadedAssetTypeEnum.None;
        public UnityEngine.Object TargetAsset { get; protected set; } = null;
        public float RemainTimeToBeDelete { get; protected set; } = 300f;
        public long MarkToDeleteTime { get; protected set; } = 0;
        public IAssetManager BelongAssetManager { get; protected set; } = null;

        public string AssetBelongBundleName { get; protected set; } //资源所属的AssetBundle 名称

        #region 构造函数& 设置
        public AssetBundleSubAssetLoadRecord()
        {

        }


        public AssetBundleSubAssetLoadRecord(string assetUrl, string assetBundleName, LoadedAssetTypeEnum typeEnum, UnityEngine.Object asset, IAssetManager manager)
        {
            Initial(assetUrl, assetBundleName, typeEnum, asset, manager);
        }

        public void Initial(string assetUrl, string assetBundleName, LoadedAssetTypeEnum typeEnum, UnityEngine.Object asset, IAssetManager manager)
        {
            Debug.Assert(manager != null);
            Debug.Assert(asset != null);

            AssetUrl = assetUrl;
            AssetBelongBundleName = assetBundleName;
            InstanceID = asset.GetInstanceID();
            ReferenceCount = 0;
            AssetLoadedType = typeEnum;
            BelongAssetManager = manager;
            TargetAsset = asset;
            RemainTimeToBeDelete = BelongAssetManager.MaxAliveTimeAfterNoReference;
        }

        #endregion


        /// <summary>
        /// 标示是否引用相同的资源
        /// </summary>
        /// <param name="record1"></param>
        /// <returns></returns>
        public bool IsReferenceSameAsset(AssetBundleSubAssetLoadRecord record1)
        {
            if (record1 != null)
            {
                if (record1.AssetUrl == this.AssetUrl && record1.AssetBelongBundleName == this.AssetBelongBundleName)
                    return true;
            }

            return false;
        }




        public void AddReference()
        {
            if (TargetAsset == null)
            {
                Debug.LogErrorFormat("资源{0} 已经被卸载了, 无法增加引用", AssetUrl);
                ReferenceCount = 0;
            }
            else
            {
                ++ReferenceCount;
            }
            NotifyReferenceChange();
        }

        public void ReduceReference(bool isforceDelete = false)
        {
            if (TargetAsset == null)
            {
                Debug.LogErrorFormat("资源{0}  已经被卸载了, 无法减少引用", AssetUrl);
                ReferenceCount = 0;
            }
            else
            {
                if (isforceDelete)
                    ReferenceCount = 0;
                else
                    --ReferenceCount;
                if (ReferenceCount == 0)
                    MarkToDeleteTime = DateTime.UtcNow.ToTimestamp_Millisecond();
            }
            NotifyReferenceChange();
        }

        public bool TimeTick(float tickTime)
        {
            if (TargetAsset == null)
            {
                ReferenceCount = 0;
                NotifyReferenceChange();
                return false;
            }

            RemainTimeToBeDelete = RemainTimeToBeDelete - tickTime;
            return RemainTimeToBeDelete > 0f;
        }


        public void NotifyNoReference()
        {
            if (TargetAsset == null) return;
            GameObject go = TargetAsset as GameObject;
            if (go != null)
            {
                go.SetActive(false);
            }
        }

        public bool NotifyReReference()
        {
            ReferenceCount = 1;
            return TargetAsset != null;
        }

        public void NotifyReleaseRecord()
        {
            BelongAssetManager.NotifyAssetRelease(this);
            AssetUrl = string.Empty;
            ReferenceCount = 0;
            AssetLoadedType = LoadedAssetTypeEnum.None;
            BelongAssetManager = null;
            TargetAsset = null;
        }

        public void NotifyReferenceChange()
        {
            BelongAssetManager.NotifyAssetReferenceChange(this);
        }


    }
}

File: GameFramePro/Assets/.lpproj/Apis/Unity/CS/GameFramePro/ResourcesEx/AssetBundleSubAssetLoadRecord.lua
---@class CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord : CS.System.Object
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord = {}

---@field public CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_AssetUrl : CS.System.String
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_AssetUrl = nil

---@field public CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_InstanceID : CS.System.Int32
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_InstanceID = nil

---@field public CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_ReferenceCount : CS.System.Int32
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_ReferenceCount = nil

---@field public CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_AssetLoadedType : CS.GameFramePro.ResourcesEx.LoadedAssetTypeEnum
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_AssetLoadedType = nil

---@field public CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_TargetAsset : CS.UnityEngine.Object
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_TargetAsset = nil

---@field public CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_RemainTimeToBeDelete : CS.System.Single
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_RemainTimeToBeDelete = nil

---@field public CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_MarkToDeleteTime : CS.System.Int64
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_MarkToDeleteTime = nil

---@field public CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_BelongAssetManager : CS.GameFramePro.ResourcesEx.IAssetManager
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_BelongAssetManager = nil

---@field public CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_AssetBelongBundleName : CS.System.String
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.Debug_AssetBelongBundleName = nil

---@property readwrite CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.InstanceID : CS.System.Int32
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.InstanceID = nil

---@property readwrite CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.AssetUrl : CS.System.String
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.AssetUrl = nil

---@property readwrite CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.ReferenceCount : CS.System.Int32
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.ReferenceCount = nil

---@property readwrite CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.AssetLoadedType : CS.GameFramePro.ResourcesEx.LoadedAssetTypeEnum
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.AssetLoadedType = nil

---@property readwrite CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.TargetAsset : CS.UnityEngine.Object
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.TargetAsset = nil

---@property readwrite CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.RemainTimeToBeDelete : CS.System.Single
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.RemainTimeToBeDelete = nil

---@property readwrite CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.MarkToDeleteTime : CS.System.Int64
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.MarkToDeleteTime = nil

---@property readwrite CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.BelongAssetManager : CS.GameFramePro.ResourcesEx.IAssetManager
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.BelongAssetManager = nil

---@property readwrite CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.AssetBelongBundleName : CS.System.String
CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord.AssetBelongBundleName = nil

---@return CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord
function CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord()
end

---@param assetUrl : CS.System.String
---@param assetBundleName : CS.System.String
---@param typeEnum : CS.GameFramePro.ResourcesEx.LoadedAssetTypeEnum
---@param asset : CS.UnityEngine.Object
---@param manager : CS.GameFramePro.ResourcesEx.IAssetManager
---@return CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord
function CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord(assetUrl, assetBundleName, typeEnum, asset, manager)
end

function CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord:UpdateData()
end

---@param assetUrl : CS.System.String
---@param assetBundleName : CS.System.String
---@param typeEnum : CS.GameFramePro.ResourcesEx.LoadedAssetTypeEnum
---@param asset : CS.UnityEngine.Object
---@param manager : CS.GameFramePro.ResourcesEx.IAssetManager
function CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord:Initial(assetUrl, assetBundleName, typeEnum, asset, manager)
end

---@param record1 : CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord
---@return CS.System.Boolean
function CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord:IsReferenceSameAsset(record1)
end

function CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord:AddReference()
end

---@param isforceDelete : CS.System.Boolean
function CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord:ReduceReference(isforceDelete)
end

---@param tickTime : CS.System.Single
---@return CS.System.Boolean
function CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord:TimeTick(tickTime)
end

function CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord:NotifyNoReference()
end

---@return CS.System.Boolean
function CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord:NotifyReReference()
end

function CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord:NotifyReleaseRecord()
end

function CS.GameFramePro.ResourcesEx.AssetBundleSubAssetLoadRecord:NotifyReferenceChange()
end